pymc/find_optim_prior.py

```diff
@@ -0,0 +1,96 @@
+#   Copyright 2021 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from typing import List
+
+import aesara
+import aesara.tensor as aet
+import numpy as np
+
+from scipy import optimize
+
+import pymc as pm
+
+__all__ = ["find_optim_prior"]
+
+
+def find_optim_prior(
+    pm_dist: pm.Distribution,
+    lower: float,
+    upper: float,
+    init_guess: List[float],
+    mass: float = 0.95,
+) -> np.ndarray:
+    """
+    Find optimal parameters to get `mass` % of probability
+    of `pm_dist` between `lower` and `upper`.
+    Note: only works for two-parameter distributions, as there
+    are exactly two constraints.
+
+    Parameters
+    ----------
+    pm_dist : pm.Distribution
+        PyMC distribution you want to set a prior on.
+        Needs to have a ``logcdf`` method implemented in PyMC.
+    lower : float
+        Lower bound to get `mass` % of probability of `pm_dist`.
+    upper : float
+        Upper bound to get `mass` % of probability of `pm_dist`.
+    init_guess: List[float]
+        Initial guess for ``scipy.optimize.least_squares`` to find the
+        optimal parameters of `pm_dist` fitting the interval constraint.
+    mass: float, default to 0.95
+        Share of the probability mass we want between ``lower`` and ``upper``.
+        Defaults to 95%.
+
+    Returns
+    -------
+    The optimized distribution parameters as a numpy array.
+    """
+    if len(pm_dist.rv_op.ndims_params) != 2:
```

Comment:
I still don't get it why you wouldn't be able to optimize a distribution with a single parameter.  There is always one parameter value that minimizes the optimization error. 

We can issue a warning if the final approximation error is too large.

For distributions with more than 3 parameters we could allow users to fix some of the parameters so that there are only 2 to optimize? For instance say we want a StudenT mu and sigma and we fix nu to a constant.

Reply:
> We can issue a warning if the final approximation error is too large.

Yeah, something like a warning that's triggered when there is +- 2% difference between the expected and the actual mass? E.g you want 95% of the mass between 1 and 2, but you got 97% (or 93%) of the mass between those bounds. That's super arbitrary of course

> For distributions with more than 3 parameters we could allow users to fix some of the parameters so that there are only 2 to optimize? For instance say we want a StudenT mu and sigma and we fix nu to a constant.

Actually, why not implement the case for 3 parameters? IIUC we just have to add another constraint, which I suggest should be the median. For distributions with more than 3 parameters, then we can allow users to fix some of the parameters, as you suggested

Reply:
We don't have closed form median expression for distributions. In general we could allow for arbitrary constraints but they might not be so intuitive.

I have no idea what the median of a half-normal distribution looks like for instance.

Reply:
I do. But I don't want to tell you.
Kidding aside, I imagine it's the same for the mean 🤔 So let's go with the two parameters with fix params when there are more

Reply:
Actually, just testing this with `Exponential`, the optimization seems to work but actually doesn't: if you ask for 95% of the mass between 0.1 and 0.4, you do get a "solution", but then you take random sample from the optimized prior and get 95% of the probability mass between... 0.04 and 4.8 😆

Reply:
I wouldn't think of optimizing the exponential with a lower bound other than zero. 

Also I guess the final alpha was also off? So it would show up with the user warning no?

Reply:
Mmmh, I think in the one-parameter case we could fix the lower bound at 0 and then just optimize based on the upper bound (that implies that the one-parameter distributions with a PyMC logcdf are the ones truncated at 0, like Exponential, HalfNormal, etc.).
I can also see that being useful is you only wnat to optimize one of two parameters, e.g I need a sigma so that Normal(1, sigma) has 95% prob mass between 0 and 2.
Does that make sense, and is it even sensible mathematically?